jctools-core/src/main/java/org/jctools/queues/atomic/SpscChunkedAtomicArrayQueue.java

```diff
@@ -15,79 +15,90 @@
 
 import org.jctools.util.Pow2;
 import org.jctools.util.RangeUtil;
-
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+import org.jctools.queues.MessagePassingQueue;
+import org.jctools.queues.MessagePassingQueue.Supplier;
+import org.jctools.queues.MessagePassingQueueUtil;
+import org.jctools.queues.QueueProgressIndicators;
+import org.jctools.queues.IndexedQueueSizeUtil;
+import static org.jctools.queues.atomic.LinkedAtomicArrayQueueUtil.*;
 import java.util.concurrent.atomic.AtomicReferenceArray;
+import org.jctools.queues.MpmcArrayQueue;
 
+/**
+ * NOTE: This class was automatically generated by org.jctools.queues.atomic.JavaParsingAtomicLinkedQueueGenerator
+ * which can found in the jctools-build module. The original source file is SpscChunkedArrayQueue.java.
+ */
 public class SpscChunkedAtomicArrayQueue<E> extends BaseSpscLinkedAtomicArrayQueue<E> {
-    private int maxQueueCapacity;
+
+    private final int maxQueueCapacity;
+
     private long producerQueueLimit;
 
-    public SpscChunkedAtomicArrayQueue(final int capacity) {
+    public SpscChunkedAtomicArrayQueue(int capacity) {
         this(Math.max(8, Pow2.roundToPowerOfTwo(capacity / 8)), capacity);
     }
 
-    public SpscChunkedAtomicArrayQueue(final int chunkSize, final int capacity) {
+    public SpscChunkedAtomicArrayQueue(int chunkSize, int capacity) {
         RangeUtil.checkGreaterThanOrEqual(capacity, 16, "capacity");
         // minimal chunk size of eight makes sure minimal lookahead step is 2
         RangeUtil.checkGreaterThanOrEqual(chunkSize, 8, "chunkSize");
-
         maxQueueCapacity = Pow2.roundToPowerOfTwo(capacity);
         int chunkCapacity = Pow2.roundToPowerOfTwo(chunkSize);
         RangeUtil.checkLessThan(chunkCapacity, maxQueueCapacity, "chunkCapacity");
-
         long mask = chunkCapacity - 1;
         // need extra element to point at next array
         AtomicReferenceArray<E> buffer = allocate(chunkCapacity + 1);
         producerBuffer = buffer;
         producerMask = mask;
         consumerBuffer = buffer;
         consumerMask = mask;
-        producerBufferLimit = mask - 1; // we know it's all empty to start with
+        // we know it's all empty to start with
+        producerBufferLimit = mask - 1;
         producerQueueLimit = maxQueueCapacity;
-        soProducerIndex(0L);// serves as a StoreStore barrier to support correct publication
```

Comment:
This volatile/ordered write at the end of construction as a substitute for final field safe publication guarantees does not work sadly (see Shipilev JMM thingy). I think this was a misguided effort in any case, queues should be safely published before used, which is generally the case anyhow.